File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/interpreter/solution/expression_parser.cpp
// expression_parser.cpp

#include "expression_builder.h"
#include "expression_parser.h"

void expression_parser::parse(expression_builder & a_expression_builder)
{
	a_expression_builder.build_variable("b");
	a_expression_builder.build_unary_minus();
	a_expression_builder.build_variable("b");
	a_expression_builder.build_variable("b");
	a_expression_builder.build_multiply();
	a_expression_builder.build_literal(4);
	a_expression_builder.build_variable("a");
	a_expression_builder.build_multiply();
	a_expression_builder.build_variable("c");
	a_expression_builder.build_multiply();
	a_expression_builder.build_binary_minus();
	a_expression_builder.build_method_invocation("sqrt", 1);
	a_expression_builder.build_binary_plus();
	a_expression_builder.build_literal(2);
	a_expression_builder.build_variable("a");
	a_expression_builder.build_multiply();
	a_expression_builder.build_divide();
}

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/null_object/lab/ExpressionParser.java
// ExpressionParser.java

public class ExpressionParser
{
	public void parse(ExpressionBuilder aExpressionBuilder)
	{
		aExpressionBuilder.buildNullExpression(); // aExpressionBuilder.buildVariable("b");
		aExpressionBuilder.buildUnaryMinus();
		aExpressionBuilder.buildVariable("b");
		aExpressionBuilder.buildVariable("b");
		aExpressionBuilder.buildMultiply();
		aExpressionBuilder.buildLiteral(4);
		aExpressionBuilder.buildVariable("a");
		aExpressionBuilder.buildMultiply();
		aExpressionBuilder.buildVariable("c");
		aExpressionBuilder.buildMultiply();
		aExpressionBuilder.buildBinaryMinus();
		aExpressionBuilder.buildMethodInvocation("sqrt", 1);
		aExpressionBuilder.buildBinaryPlus();
		aExpressionBuilder.buildLiteral(2);
		aExpressionBuilder.buildVariable("a");
		aExpressionBuilder.buildMultiply();
		aExpressionBuilder.buildDivide();
	}
}
